src/utils/genetic_operators.cpp

```diff
@@ -0,0 +1,203 @@
+{
+
+// Implementation of the binary crossover.
+// Requires the distribution index eta_c in [1, 100], crossover probability p_cr  in[0,1] -> undefined algo behaviour
+// otherwise Requires dimensions of the parent and bounds to be equal -> out of bound reads. dim_i is the integer
+// dimension (integer alleles assumed at the end of the chromosome)
+
+std::pair<vector_double, vector_double> sbx_crossover(const vector_double &parent1, const vector_double &parent2,
+                                                      const std::pair<vector_double, vector_double> &bounds,
+                                                      vector_double::size_type dim_i, const double p_cr,
+                                                      const double eta_c, detail::random_engine_type &random_engine)
+{
+    // Decision vector dimensions
+    auto dim = parent1.size();
+    auto dim_c = dim - dim_i;
+    // Problem bounds
+    const vector_double &lb = bounds.first;
+    const vector_double &ub = bounds.second;
+    // declarations
+    double y1, y2, yl, yu, rand01, beta, alpha, betaq, c1, c2;
+    vector_double::size_type site1, site2;
+    // Initialize the child decision vectors
+    vector_double child1 = parent1;
+    vector_double child2 = parent2;
+    // Random distributions
+    std::uniform_real_distribution<> drng(0., 1.); // to generate a number in [0, 1)
+
+    // This implements a Simulated Binary Crossover SBX and applies it to the non integer part of the decision
+    // vector
+    if (drng(random_engine) < p_cr) {
+        for (decltype(dim_c) i = 0u; i < dim_c; i++) {
+            if ((drng(random_engine) <= 0.5) && (std::abs(parent1[i] - parent2[i])) > 1e-14 && lb[i] != ub[i]) {
+                if (parent1[i] < parent2[i]) {
+                    y1 = parent1[i];
+                    y2 = parent2[i];
+                } else {
+                    y1 = parent2[i];
+                    y2 = parent1[i];
+                }
+                yl = lb[i];
+                yu = ub[i];
+                rand01 = drng(random_engine);
+                beta = 1. + (2. * (y1 - yl) / (y2 - y1));
+                alpha = 2. - std::pow(beta, -(eta_c + 1.));
+                if (rand01 <= (1. / alpha)) {
+                    betaq = std::pow((rand01 * alpha), (1. / (eta_c + 1.)));
+                } else {
+                    betaq = std::pow((1. / (2. - rand01 * alpha)), (1. / (eta_c + 1.)));
+                }
+                c1 = 0.5 * ((y1 + y2) - betaq * (y2 - y1));
+                beta = 1. + (2. * (yu - y2) / (y2 - y1));
+                alpha = 2. - std::pow(beta, -(eta_c + 1.));
+                if (rand01 <= (1. / alpha)) {
+                    betaq = std::pow((rand01 * alpha), (1. / (eta_c + 1.)));
+                } else {
+                    betaq = std::pow((1. / (2. - rand01 * alpha)), (1. / (eta_c + 1.)));
+                }
+                c2 = 0.5 * ((y1 + y2) + betaq * (y2 - y1));
+                if (c1 < lb[i]) c1 = lb[i];
```

Comment:
Note that c++17 has std::clamp: https://en.cppreference.com/w/cpp/algorithm/clamp
which is probably optimized.

 You may wish to add your own implementation of clamp and replace it with the standard once you require C++17.

There are several places in this file where you can use clamp().

Reply:
Yeps. This code is legacy from third party and I would rather not touch it too much.